text/0431-sagemaker-l2-endpoint.md

```diff
@@ -0,0 +1,1150 @@
+
+    constructor(scope: Construct, id: string, props: ModelProps) { ... }
+  }
+  ```
+
+##### Container Definition
+
+When defining a model above, the `ContainerDefinition` interface encapsulates both the specification
+of model inference code as a `ContainerImage` and an optional set of artifacts as `ModelData`. The
+image is specified as a Docker registry path while the model artifacts must be stored in S3.
+
+- `ContainerDefinition` -- describes the container, as part of model definition above
+
+  ```ts
+  export interface ContainerDefinition {
+    /**
+     * The image used to start a container.
+     */
+    readonly image: ContainerImage;
+
+    /**
+     * A map of environment variables to pass into the container.
+     *
+     * @default none
+     */
+    readonly environment?: {[key: string]: string};
+
+    /**
+     * Hostname of the container.
+     *
+     * @default none
+     */
+    readonly containerHostname?: string;
+
+    /**
+     * S3 path to the model artifacts.
+     *
+     * @default none
+     */
+    readonly modelData?: ModelData;
+  }
+  ```
+
+###### Container Image
+
+The following interface and abstract class provide mechanisms for configuring a container image.
+These closely mirror [analogous entities from the `@aws-cdk/ecs` module][ecs-image] but, rather than
+`bind`-ing upon an ECS task definition, instead operate upon a SageMaker model.
+
+[ecs-image]: https://github.com/aws/aws-cdk/blob/572b52c45a9eb08b62a0f9cc6520c1722089bae6/packages/@aws-cdk/aws-ecs/lib/container-image.ts
+
+- `ContainerImageConfig` -- the configuration for creating a container image
+
+  ```ts
+  export interface ContainerImageConfig {
+    /**
+     * The image name. Images in Amazon ECR repositories can be specified by either using the full registry/repository:tag or
+     * registry/repository@digest.
+     *
+     * For example, 012345678910.dkr.ecr.<region-name>.amazonaws.com/<repository-name>:latest or
+     * 012345678910.dkr.ecr.<region-name>.amazonaws.com/<repository-name>@sha256:94afd1f2e64d908bc90dbca0035a5b567EXAMPLE.
+     */
+    readonly imageName: string;
+  }
+  ```
+
+- `ContainerImage` -- abstract class defining `bind` contract for images alongside static factory
+  methods to enable different sources (e.g., image in ECR repository, local Dockerfile)
+
+  ```ts
+  export abstract class ContainerImage {
+    /**
+     * Reference an image in an ECR repository
+     */
+    public static fromEcrRepository(repository: ecr.IRepository, tag: string = 'latest'): ContainerImage { ... }
+
+    /**
+     * Reference an image that's constructed directly from sources on disk
+     *
+     * @param scope The scope within which to create the image asset
+     * @param id The id to assign to the image asset
+     * @param props The properties of a Docker image asset
+     */
+    public static fromAsset(scope: Construct, id: string, props: assets.DockerImageAssetProps): ContainerImage { ... }
```

Comment:
@rix0rrr, as I was implementing this change, I noticed that the [ECS module's `AssetImage` hardcodes the construct ID as `'AssetImage'`](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/aws-ecs/lib/images/asset-image.ts#L26). Bringing this updated API as-is into the SageMaker module would present a rough edge to customers as it would prevent customers from using the API to define inference pipelines (represented as multiple containers), the reason being that the second `bind` invocation (for the second container) would attempt to reuse a hardcoded construct ID resulting in a an error like:
```
Error: There is already a Construct with name 'AssetImage' in Model [InferencePipelineModel]
```

I see that [Elad introduced into the ECS module the API `ContainerImage.fromDockerImageAsset`](https://github.com/aws/aws-cdk/commit/b94577a0acb0aa76e257fb15267faf6c0954700a) which operates on an `DockerImageAsset` (the construct ID for which the customer has full control).

To sidestep the rough edge for customers, I'm considering replacing `ContainerImage.fromAsset` with `ContainerImage.fromDockerImageAsset`. Does that seem reasonable or would you prefer I keep both APIs?

Reply:
In rev2, I've decided to pivot to having the new proposed APIs take an `Asset` from '@aws-cdk/aws-s3-assets' (for `ModelData`) and `DockerImageAsset` from '@aws-cdk/aws-ecr-assets' (for `ContainerImage`) to avoid adding any assets into the hierarchy myself.